Mathlib/Tactic/Widget/StringDiagram.lean

```diff
@@ -0,0 +1,331 @@
+def addInstruction (i : String) : DiagramBuilderM Unit := do
+  modify fun st => { st with sub := st.sub ++ s!"{i}\n" }
+
+/-- Add constructor `tp v := nm (vs)` to the substance program. -/
+def addConstructor (tp : String) (v : PenroseVar) (nm : String) (vs : List PenroseVar) :
+    DiagramBuilderM Unit := do
+  let vs' := ", ".intercalate (vs.map (fun v => toString v))
+  addInstruction s!"{tp} {v} := {nm} ({vs'})"
+
+/-- Run the program in the diagram builder monad. -/
+def DiagramBuilderM.run {α : Type} (x : DiagramBuilderM α) : MonoidalM α :=
+  x.run' {}
+
+open scoped Jsx in
+/-- Construct a string diagram from a Penrose `sub`stance program and expressions `embeds` to
+display as labels in the diagram. -/
+def mkStringDiag (e : Expr) : MonoidalM Html := do
+  DiagramBuilderM.run do
+    let e' ← eval e
+    let nodes ← e'.nodes
+    let strands ← e'.strands
+    /- Add 2-morphisms. -/
+    for x in nodes.join do
+      match x with
+      | .atom _ => do addPenroseVar "Atom" x.toPenroseVar
+      | .id _ => do addPenroseVar "Id" x.toPenroseVar
+    /- Add constraints. -/
+    for l in nodes do
+      for (x₁, x₂) in pairs l do
+        addInstruction s!"Left({x₁.toPenroseVar}, {x₂.toPenroseVar})"
+    /- Add constraints. -/
+    for (l₁, l₂) in pairs nodes do
+      if let .some x₁ := l₁.head? then
+        if let .some x₂ := l₂.head? then
+          addInstruction s!"Above({x₁.toPenroseVar}, {x₂.toPenroseVar})"
+    /- Add 1-morphisms as strings. -/
+    for l in strands do
+      for s in l do
+        addConstructor "Mor1" s.toPenroseVar
+          "MakeString" [s.startPoint.toPenroseVar, s.endPoint.toPenroseVar]
+    match ← buildDiagram with
+    | some html => return html
+    | none => return <span>No 2-morphisms.</span>
+
+/-- Given a 2-morphism, return a string diagram. Otherwise `none`. -/
+def stringM? (e : Expr) : MetaM (Option Html) := do
+  let e ← instantiateMVars e
+  return some <| ← MonoidalM.run (← mkContext e) <| mkStringDiag e
+
+/-- Given an equality between 2-morphisms, return a string diagram of the LHS. Otherwise `none`. -/
+def stringLeftM? (e : Expr) : MetaM (Option Html) := do
+  let e ← instantiateMVars e
+  let some (_, lhs, _) := e.eq? | return none
+    return some <| ← MonoidalM.run (← mkContext lhs) <| mkStringDiag lhs
+
+/-- Given an equality between 2-morphisms, return a string diagram of the RHS. Otherwise `none`. -/
+def stringRightM? (e : Expr) : MetaM (Option Html) := do
+  let e ← instantiateMVars e
+  let some (_, _, rhs) := e.eq? | return none
+    return some <| ← MonoidalM.run (← mkContext rhs) <| mkStringDiag rhs
+
+/-- The string diagram widget. -/
+@[expr_presenter]
+def stringPresenter : ExprPresenter where
+  userName := "String diagram"
+  layoutKind := .block
+  present type := do
+    if let some d ← stringM? type then
+      return d
+    throwError "Couldn't find a string diagram."
+
+/-- The string diagram widget. -/
+@[expr_presenter]
```

Comment:
Instead of having two presenters, how about having just one that shows both sides? That way it's easy to see the difference and what needs to be done to match the two sides up. You could do it like this:
```lean
@[expr_presenter]
def stringPresenter : ExprPresenter where
  userName := "Equality of string diagrams"
  layoutKind := .block
  present type := do
    let some d ← stringLeftM? type
      | throwError "Couldn't find a string diagram."
    let some e ← stringRightM? type
      | throwError "Couldn't find a string diagram."
    return <div className="flex">
      <div className="w-50">{d}</div>
      <div className="w-50">{e}</div>
    </div>
```